### ISSUE TITLE
Replicate underlying objects to a user-provided S3 bucket

### ISSUE THREAD

**Issue by inian** [supabase/storage-api#27]
We should make it easy for users to export their objects. Also useful for backups etc.  We can probably use the [S3 replication functionality](https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication.html) to do this. 

Relevant discussion https://github.com/supabase/supabase/discussions/1930

**Comment by d-mok** [comment]
Is there any update on this? The lack of backup-from-supabase-storage function is stopping me to use supabase storage.

**Comment by inian** [comment]
Hi @d-mok, we couldn't find an easy way of doing this with the S3 replication functionality above since there there is a [limit](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3api/put-bucket-replication.html) of 1,000 replication policies per bucket which doesn't work at our scale. 

If you have any ideas on how to go about implementing this in a multi-tenant way, let me know!

**Comment by ornithophile** [comment]
@inian Hi, are there any other plans for syncing files out of Supabase Storage?

**Comment by d-mok** [comment]
@inian 
Then, may be the direction to go is to allow users to provide their own S3 bucket? Like, a bring-your-own-bucket approach?

**Comment by ornithophile** [comment]
This would be perfect.

**Comment by thebengeu** [comment]
What I'm thinking is to have a new replication section in the UI where you can enter an access key ID and secret access key for an AWS IAM user, and a bucket, and then we'll regularly replicate to that bucket. How does that sound?

**Comment by d-mok** [comment]
> What I'm thinking is to have a new replication section in the UI where you can enter an access key ID and secret access key for an AWS IAM user, and a bucket, and then we'll regularly replicate to that bucket. How does that sound?

So we (users) provide a S3 bucket, and supabase regularly backup everything to that bucket? That sounds good enough.
But I think a better option is to allow user to bring-their-own-bucket. In this way, I can interact with my bucket both through supabase or via some other s3 client. It would be best if, for example, when I add a file to my bucket through another client, supabase can also register the change. This is consistent with the principle how supabase is dealing with Postgresql: we can use the postgresql database any way we want, and supabase will work accordingly.

**Comment by tehnorm** [comment]
I'd mirror what @d-mok is expressing. Bring your own S3 bucket is very useful for cases that data governance is a consideration as well. Usually for larger organizations have these challenges. It seems lately many SaaS offers do provide a bring-your-own-bucket model.

**Comment by grschafer** [comment]
Another vote in favor of bring-your-own-bucket: I think it would allow us to manage data transfer costs better (e.g. keeping transfers within AWS infrastructure in the same region and AZ) which is important for data/processing-heavy applications.

I opened a separate discussion about same-region transfer costs (https://github.com/supabase/supabase/discussions/3403) to not distract from the replication focus of this issue, but I wanted to leave a comment here and link the two ideas, because bring-your-own-bucket seems to be a good solution for both.

**Comment by kro-dev** [comment]
I’d also like to see BYOB!

**Comment by rlee1990** [comment]
Any update on this?

**Comment by narve** [comment]
Yes, status please? Is it coming (soon), or should we find other workarounds?

**Comment by bkuermayr** [comment]
would also love to see this

**Comment by terehov** [comment]
Any updates. For a professional service, I need to be able to back up everything myself and prove that. 
Without the ability to bring my own storage, and without a fully compatible S3 interface, there is no way of doing so. 

PS: Not even RLS is exported into migrations and seeds?

**Comment by sapoepsilon** [comment]
Any updates?

**Comment by inian** [comment]
I am trying to better understand the usecases here. 

Can you share which of the following problems you are looking to solve?

1. compliance - your security posture requires your storage objects to be within your own cloud
2. missing feature backups
3. missing feature cross region replication - for higher availability 
4. lower storage costs 
	a. use your own AWS account to make use of any discounted pricing you have
    b. use a S3 compatible storage service which provides lower cost / GB (like Wasabi)
5. need an S3 compatible interface for my tools ([supported now](https://supabase.com/blog/s3-compatible-storage))
6. lower egress costs if the S3 bucket is within the same VPC as your other workloads
7. anything else i am missing?

In case of missing features like cross region replication or backups, would you still need BYOB if this was natively supported in the storage server?

**Comment by narve** [comment]
Speaking for myself, my use case was backup - but this could have been fixed by now, I haven't kept up2date on changes lately. 

But I think you will find that all of the points you mentioned are relevant. Also point 1 "Complicance" could include GDPR considerations.

**Comment by IPJT** [comment]
@inian -
Hi there,
I'd like to backup the storage files! 
Thanks :)

**Comment by limegorilla** [comment]
For me I need to be able to prove that I have **full and complete** backups, and can recover from a complete system failure.

Essentially my DR should be able to handle Supabase Cloud going completely offline, and being able to restore from a recent (max 24h) backup.

**Comment by deduble** [comment]
how is this coming along. or were you meaning to just use S3 wrapper from integrations?
